File: server/ml/faceDetection/decoder.js
const STRIDES = [8, 16, 32];
const NUM_ANCHORS = 2;
const INPUT_SIZE = 640;

const anchorCache = {};

function generateAnchors(stride) {
  if (anchorCache[stride]) return anchorCache[stride];

  const featureSize = INPUT_SIZE / stride;

  const anchors = [];

  for (let y = 0; y < featureSize; y++) {
    for (let x = 0; x < featureSize; x++) {
      const cx = (x + 0.5) * stride;
      const cy = (y + 0.5) * stride;

      // SCRFD 2-anchor model
      for (let a = 0; a < NUM_ANCHORS; a++) {
        anchors.push({
          x: cx,
          y: cy,
        });
      }
    }
  }

  anchorCache[stride] = anchors;
  return anchors;
}

function decodeBox(anchor, bbox, stride) {
  const left = bbox[0] * stride;
  const top = bbox[1] * stride;
  const right = bbox[2] * stride;
  const bottom = bbox[3] * stride;

  return {
    x1: anchor.x - left,
    y1: anchor.y - top,
    x2: anchor.x + right,
    y2: anchor.y + bottom,
    width: left + right,
    height: top + bottom,
  };
}

function decodeLandmarks(anchor, kps, stride) {
  const landmarks = [];

  for (let i = 0; i < 5; i++) {
    landmarks.push({
      x: anchor.x + kps[i * 2] * stride,
      y: anchor.y + kps[i * 2 + 1] * stride,
    });
  }

  return landmarks;
}

export function decodeSCRFD(outputs, threshold = 0.5) {
  const detections = [];

  for (const stride of STRIDES) {
    const scores = outputs[`score_${stride}`].data;
    const boxes = outputs[`bbox_${stride}`].data;
    const kps = outputs[`kps_${stride}`].data;

    const anchors = generateAnchors(stride);

    for (let i = 0; i < scores.length; i++) {
      const score = scores[i];

      if (score < threshold) continue;

      const box = decodeBox(
        anchors[i],
        boxes.subarray(i * 4, i * 4 + 4),
        stride
      );

      const landmarks = decodeLandmarks(
        anchors[i],
        kps.subarray(i * 10, i * 10 + 10),
        stride
      );

      detections.push({
        score,
        box,
        landmarks,
      });
    }
  }

  return detections;
}

File: ml-server/skills/face_detection/decoder.py
import numpy as np

STRIDES = [8, 16, 32]
NUM_ANCHORS = 2
INPUT_SIZE = 640

anchor_cache = {}


def generate_anchors(stride):

    if stride in anchor_cache:
        return anchor_cache[stride]

    feature_size = INPUT_SIZE // stride

    anchors = []

    for y in range(feature_size):
        for x in range(feature_size):

            cx = x * stride
            cy = y * stride

            for _ in range(NUM_ANCHORS):

                anchors.append({
                    "x": cx,
                    "y": cy
                })

    anchor_cache[stride] = anchors

    return anchors


def decode_box(anchor, bbox, stride):
    left = bbox[0] * stride
    top = bbox[1] * stride
    right = bbox[2] * stride
    bottom = bbox[3] * stride

    return {
        "x1": float(anchor["x"] - left),
        "y1": float(anchor["y"] - top),
        "x2": float(anchor["x"] + right),
        "y2": float(anchor["y"] + bottom),
        "width": float(left + right),
        "height": float(top + bottom),
    }


def decode_landmarks(anchor, kps, stride):
    landmarks = []
    for i in range(5):
        landmarks.append({
            "x": float(anchor["x"] + kps[i * 2] * stride),
            "y": float(anchor["y"] + kps[i * 2 + 1] * stride),
        })
    return landmarks


def decode_scrfd(outputs, original_width, original_height, scale, threshold=0.5):
    output_map = {
        "score_8": outputs[0],
        "score_16": outputs[1],
        "score_32": outputs[2],
        "bbox_8": outputs[3],
        "bbox_16": outputs[4],
        "bbox_32": outputs[5],
        "kps_8": outputs[6],
        "kps_16": outputs[7],
        "kps_32": outputs[8],
    }

    detections = []

    for stride in STRIDES:
        scores = output_map[f"score_{stride}"].reshape(-1)
        boxes = output_map[f"bbox_{stride}"].reshape(-1, 4)
        kps = output_map[f"kps_{stride}"].reshape(-1, 10)

        anchors = generate_anchors(stride)

        for i in range(len(scores)):
            score = float(scores[i])
            if score < threshold:
                continue

            box = decode_box(anchors[i], boxes[i], stride)
            landmarks = decode_landmarks(anchors[i], kps[i], stride)

            # Undo the letterbox scale (single uniform factor — no distortion)
            box["x1"] /= scale
            box["x2"] /= scale
            box["width"] /= scale
            box["y1"] /= scale
            box["y2"] /= scale
            box["height"] /= scale

            for p in landmarks:
                p["x"] /= scale
                p["y"] /= scale

            # Clamp to original image bounds
            box["x1"] = max(0, min(box["x1"], original_width))
            box["x2"] = max(0, min(box["x2"], original_width))
            box["y1"] = max(0, min(box["y1"], original_height))
            box["y2"] = max(0, min(box["y2"], original_height))

            detections.append({
                "score": score,
                "box": box,
                "landmarks": landmarks,
            })

    return detections